### ISSUE TITLE
MatSlider: change event fired twice

### ISSUE THREAD

**Issue by PhilippS93** [angular/components#14363]
#### Bug:
MatSlider change event is fired twice when slider knob is dragged and released. Surprisingly enough, this happens only sometimes.

#### What is the expected behavior?
Event emitted when the slider value has changed.

#### What is the current behavior?
MatSlider change event is fired twice when slider knob is dragged and released.

#### What are the steps to reproduce?
See: [stackblitz](https://stackblitz.com/edit/angular-mcbka8?file=app%2Fslider-overview-example.ts)

**Comment by manklu** [comment]
It fires immediately if you don't click exactly in the center of the knob.

**Comment by crisbeto** [comment]
It probably because we change the value on `mousedown` and then while the user is dragging. This sounds like it's working as expected.

**Comment by cmedinadeveloper** [comment]
This is happening to me on every single iOS devices, any suggestions?

**Comment by notbenj** [comment]
> It probably because we change the value on `mousedown` and then while the user is dragging. This sounds like it's working as expected.

I'm sure an end-user doesn't expect the value to be changed, just because he/she grabbed the knob 1 pixel on the left of the knob center :)

I think the behavior should be on the mouseup event, to avoid this situation.

**Comment by crisbeto** [comment]
The problem, if we were to do that, is that the events will no longer be in sync with what is being rendered on screen.

**Comment by Domvel** [comment]
Confirmed, it's an unexpected behavior (bug).
https://stackblitz.com/edit/angular-mat-slider-full-options?embed=1&file=app/slider-configurable-example.html

If you click on the thumb if the slider is not focused an begin to drag the change event is emitted. 
Also if you slick on the range (line, not the thumb) the change event is always fired on mouse button down. And final on mouse up (as expected).

**Comment by Domvel** [comment]
Here's the current behavior. 
![export](https://user-images.githubusercontent.com/35452565/54692756-cdcabd80-4b25-11e9-9249-85c37f86998d.gif)
You see that the change event is fired when the user just press the mouse button down to begin dragging. This is an unexpected behavior. (bug) It's expected on mouse button up.

**Comment by Domvel** [comment]
Workaround: 
Use the event `slideend`.

```html
<mat-slider
  #mySlider
  (slideend)="sliderOnChange(mySlider.value)"
></mat-slider>
````

This is the "real" (expected) "change"-event.

**Comment by BorisZubchenko** [comment]
Well, the `slideend` event is not available now :(

**Comment by sksuri100** [comment]
> #### Bug:
> MatSlider change event is fired twice when slider knob is dragged and released. Surprisingly enough, this happens only sometimes.
> 
> #### What is the expected behavior?
> Event emitted when the slider value has changed.
> 
> #### What is the current behavior?
> MatSlider change event is fired twice when slider knob is dragged and released.
> 
> #### What are the steps to reproduce?
> See: [stackblitz](https://stackblitz.com/edit/angular-mcbka8?file=app%2Fslider-overview-example.ts)
> 
> #### What is the use-case or motivation for changing an existing behavior?
> #### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
> #### Is there anything else we should know?

i have faced same problem please let me know what is right solution of this bug????

**Comment by Sn3b** [comment]
> Well, the `slideend` event is not available now :(

I have a little work around that seems to work so far... Please don't fire at me if it doesn't for you ;)

```html
<mat-slider min="0"
            max="100"
            step="1"
            (input)="sliderSliding( $event )"
            (change)="sliderChanged( $event )"
            (mouseup)="sliderMouseButtonReleased( $event )"
            [(ngModel)]="value">
</mat-slider>

```

```typescript
    public value: number = 0;
    private _slidingValue: number;

    public sliderSliding( args: MatSliderChange ): void {
        this._slidingValue = args.value;
    }

    public sliderMouseButtonReleased( args: MouseEvent ) {
        if ( this.value === this._slidingValue ) {
            this.sliderChanged( null );
        }
    }

    public sliderChanged( args: any ): void {
        if ( !!args && !!args.source._lastPointerEvent ) {
            return;
        }

        // Your code to process the event here
    }

```

**Comment by KrzysztofMilewski** [comment]
For those who - like me - desperately need a workaround: I downloaded the material repository and moved all necessary files to my project. Then, in src/material/slider/slider.ts in method `_pointerDown` (line 620) I removed line 644, which is this: `this._emitChangeEvent();`.

This works for my case, but bear in mind, that I haven't run any automated tests againist this change. Hope this helps

**Comment by samworley** [comment]
> > Well, the `slideend` event is not available now :(
> 
> I have a little work around that seems to work so far... Please don't fire at me if it doesn't for you ;)
> 
> ```
> <mat-slider min="0"
>             max="100"
>             step="1"
>             (input)="sliderSliding( $event )"
>             (change)="sliderChanged( $event )"
>             (mouseup)="sliderMouseButtonReleased( $event )"
>             [(ngModel)]="value">
> </mat-slider>
> ```
> 
> ```ts
>     public value: number = 0;
>     private _slidingValue: number;
> 
>     public sliderSliding( args: MatSliderChange ): void {
>         this._slidingValue = args.value;
>     }
> 
>     public sliderMouseButtonReleased( args: MouseEvent ): void {
>         if ( this.value === this._slidingValue ) {
>             this.sliderChanged( null );
>         }
>     }
> 
>     public sliderChanged( args: any ): void {
>         if ( !!args && !!args.source._lastPointerEvent ) {
>             return;
>         }
> 
>         // Your code to process the event here
>     }
> ```

Thanks! This helped me out a lot, but I noticed that you will still get multiple events fired when using a touch device or if using a keyboard to move the slider.
I found a simpler way of working around the issue is instead of binding to the change event, to bind to mouseup, keyup and touchend to cover keyboard, mouse and touch. If you do this you no longer need the sliderMouseButtonReleased function and can remove the return at the start of sliderChanged e.g.

```html
<mat-slider min="0"
            max="100"
            step="1"
            (input)="sliderSliding( $event )"
            (mouseup)="sliderChanged( $event )"
            (keyup)="sliderChanged( $event )"
            (touchend)="sliderChanged( $event )"
            [(ngModel)]="value">
</mat-slider>
```

```ts
    public value: number = 0;
    private _slidingValue: number;

    public sliderSliding( args: MatSliderChange ): void {
        this._slidingValue = args.value;
    }

    public sliderChanged( args: any ): void {
        // Your code to process the event here
    }
```

**Comment by vanrado** [comment]
> ```
> (mouseup)="sliderChanged( $event )"
>             (keyup)="sliderChanged( $event )"
>             (touchend)="sliderChanged( $event )"
> ```

touchend event is not emitted by mat-slider. Did you test that pls?

**Comment by samworley** [comment]
> > ```
> > (mouseup)="sliderChanged( $event )"
> >             (keyup)="sliderChanged( $event )"
> >             (touchend)="sliderChanged( $event )"
> > ```
> 
> touchend event is not emitted by mat-slider. Did you test that pls?

Appeared to work when I tested it - without touchend and after removing changes was not getting any events fired from a touch device.

**Comment by vanrado** [comment]
> > > Well, the `slideend` event is not available now :(
> > 
> > 
> > I have a little work around that seems to work so far... Please don't fire at me if it doesn't for you ;)
> > ```
> > <mat-slider min="0"
> >             max="100"
> >             step="1"
> >             (input)="sliderSliding( $event )"
> >             (change)="sliderChanged( $event )"
> >             (mouseup)="sliderMouseButtonReleased( $event )"
> >             [(ngModel)]="value">
> > </mat-slider>
> > ```
> > 
> > 
> > ```ts
> >     public value: number = 0;
> >     private _slidingValue: number;
> > 
> >     public sliderSliding( args: MatSliderChange ): void {
> >         this._slidingValue = args.value;
> >     }
> > 
> >     public sliderMouseButtonReleased( args: MouseEvent ): void {
> >         if ( this.value === this._slidingValue ) {
> >             this.sliderChanged( null );
> >         }
> >     }
> > 
> >     public sliderChanged( args: any ): void {
> >         if ( !!args && !!args.source._lastPointerEvent ) {
> >             return;
> >         }
> > 
> >         // Your code to process the event here
> >     }
> > ```
> 
> Thanks! This helped me out a lot, but I noticed that you will still get multiple events fired when using a touch device or if using a keyboard to move the slider.
> I found a simpler way of working around the issue is instead of binding to the change event, to bind to mouseup, keyup and touchend to cover keyboard, mouse and touch. If you do this you no longer need the sliderMouseButtonReleased function and can remove the return at the start of sliderChanged e.g.
> 
> ```
> <mat-slider min="0"
>             max="100"
>             step="1"
>             (input)="sliderSliding( $event )"
>             (mouseup)="sliderChanged( $event )"
>             (keyup)="sliderChanged( $event )"
>             (touchend)="sliderChanged( $event )"
>             [(ngModel)]="value">
> </mat-slider>
> ```
> 
> ```ts
>     public value: number = 0;
>     private _slidingValue: number;
> 
>     public sliderSliding( args: MatSliderChange ): void {
>         this._slidingValue = args.value;
>     }
> 
>     public sliderChanged( args: any ): void {
>         // Your code to process the event here
>     }
> ```

pls

> > > ```
> > > (mouseup)="sliderChanged( $event )"
> > >             (keyup)="sliderChanged( $event )"
> > >             (touchend)="sliderChanged( $event )"
> > > ```
> > 
> > 
> > touchend event is not emitted by mat-slider. Did you test that pls?
> 
> Appeared to work when I tested it - without touchend and after removing change was not getting any events fired from a touch device.

touchend doesn't work for me. I used panend (from Hammer.js) event instead of it and it work properly now. Thank you for sharing workaround

**Comment by crisbeto** [comment]
Bumping to P2 since people seem to continue bumping into this. I also submitted a fix at #20240, but I suspect that it might be difficult to get it merged in, because it can break people's tests.

**Comment by samworley** [comment]
Thanks for fixing! Workaround no longer required
